chain/db.go

```diff
@@ -1026,3 +1026,67 @@ func NewDBStore(db DB, n *consensus.Network, genesisBlock types.Block, logger Mi
 	cs, _ := dbs.State(index.ID)
 	return dbs, cs, err
 }
+
+// NewDBStoreAtCheckpoint creates a new DBStore initialized at the provided
+// checkpoint. The checkpoint must be a v2 block. The DB will be automatically
+// migrated if necessary. The provided logger may be nil.
+func NewDBStoreAtCheckpoint(db DB, cs consensus.State, b types.Block, logger MigrationLogger) (_ *DBStore, _ consensus.State, err error) {
```

Comment:
Would it be possible to instead use `NewDBStore` as it is right now and then add a `ApplyCheckpoint` method? To make sure if you are handed a checkpoint by a third party that doesn't match your expected genesis block, you will notice? It's not obvious to me why we wouldn't apply the genesis block for that additional consistency guarantee.

Reply:
Peers don't send you the `consensus.Network`; you supply that yourself. https://github.com/SiaFoundation/coreutils/blob/b0b67c94c698f04e19f0392454aa8ff76180c2ca/syncer/peer.go#L172

I considered a method on `DBStore`, but then you have to deal with `ApplyCheckpoint` being called on a DB that's already been used; what do you do? There is no real way to handle this, so you'd just have to panic or return an error. At that point I'd rather have a separate constructor.